internal/certmonitor/certmonitor.go

```
@@ -0,0 +1,214 @@
+		return fmt.Errorf("timed out waiting for caches to sync")
+	}
+	return nil
+}
+
+// NewCertMonitor creates new instance of certMonitor
+func NewCertMonitor(config *Config, informerFactory informers.SharedInformerFactory, secretInformer cache.SharedIndexInformer, namespaceGetter NamespaceGetter) *certMonitor {
+	return &certMonitor{
+		informerfactory: informerFactory,
+		secretInformer:  secretInformer, // pragma: allowlist secret
+		config:          config,
+		namespaceGetter: namespaceGetter,
+		metrics:         fleetshardmetrics.MetricsInstance(),
+	}
+}
+
+// objectMatchesSelector checks if object matches given label selector
+func objectMatchesSelector(obj runtime.Object, selector *metav1.LabelSelector) bool {
+	if selector == nil {
+		return true
+	}
+	labelselector, err := metav1.LabelSelectorAsSelector(selector)
+	if err != nil {
+		return false
+	}
+
+	metaObj, ok := obj.(metav1.Object)
+	if !ok {
+		return false
+	}
+
+	return labelselector.Matches(labels.Set(metaObj.GetLabels()))
+
+}
+
+// secretMatches checks if a secret matches a monitor config
+func (c *certMonitor) secretMatches(s *corev1.Secret, monitor MonitorConfig) bool {
+	if s == nil {
+		return false
+	}
+	if len(monitor.Secret.Name) > 0 && s.Name != monitor.Secret.Name {
+		return false
+	}
+	if len(monitor.Namespace.Name) > 0 && s.Namespace != monitor.Namespace.Name {
+		return false
+	}
+	if monitor.Secret.LabelSelector != nil && !objectMatchesSelector(s, monitor.Secret.LabelSelector) {
+		return false
+	}
+
+	if monitor.Namespace.LabelSelector != nil {
+		ns, err := c.namespaceGetter.Get(s.Namespace)
+		if err != nil {
+			return false
+		}
+		if !objectMatchesSelector(ns, monitor.Secret.LabelSelector) {
+			return false
+		}
+	}
+	return true
+}
+
+// processSecret extracts, decodes, parses certificates from a secret, and populates prometheus metrics
+func (c *certMonitor) processSecret(secret *corev1.Secret) {
+	for dataKey, dataCert := range secret.Data {
```

Not every `dataKey` is a certificate. For example, the cert key is not a certificate.
I guess it's safe to rely on the secret structure here, e.g. explicitly get `ca.pem` and `cert.pem` from `secret.Data`

---

This was intentional, e.g. to fail silently & grab everything that is a cert - because we might monitor other types of certificates in the future ?